### PR TITLE
Custom actions/1291 teleportation as a custom action

### LINKED ISSUE #1291: Implement Chiventure example of player teleportation with custom actions
Player teleportation has been toyed with my previous action management teams. One promising solution is to leave teleportation as an action implemented using custom actions. We will create an example of using custom actions to do player teleportation to test its feasibility.

### PR DESCRIPTION
**Pull request by jbrouwer5**
We implemented TELEPORTATION as a custom action. A developer can now create a custom action that moves a player to a specified destination room. This involved creating a parameter struct that the action_new family of functions takes. This allows for more types of parameters rather than just attributes. We also implemented the applicable tests in the tests directory.

### CODE REVIEW

File: src/custom-actions/include/action_block.h [hunk 1]
```diff
@@ -8,88 +8,99 @@
...
-/* 
- * Initializes an action block. 
- * 
- * Parameters: 
- * - action block. Must point to already allocated memory.  
+/*
+ * Initializes an action block.
+ *
+ * Parameters:
+ * - action block. Must point to already allocated memory.
```
Comment (sabinecodes15) [inline src/custom-actions/include/action_block.h:82]:
```suggestion
 * - action block. Must already be allocated in memory.
```

File: src/custom-actions/src/action_block.c [hunk 2]
```diff
@@ -28,25 +29,25 @@ action_block_t* action_block_new(action_enum_t action_type, int num_args,
...
+                                  action_param_t *action_params)
 {
     AST_block_t *ast;
     action_block_t *action;
     int new_action;
     block_type_t block_type = ACTION;
-    
+
     action = malloc(sizeof(action_block_t));
     ast = malloc(sizeof(AST_block_t));
```
Comment (sabinecodes15) [inline src/custom-actions/src/action_block.c:52]: You check if `action` was properly malloced in the if statement below. Should you not also check if `ast` was properly allocated?

File: src/custom-actions/src/action_block.c [hunk 3]
```diff
@@ -56,30 +57,30 @@ AST_block_t* AST_action_block_new(action_enum_t action_type, int num_args,
...
+    if (new_action != SUCCESS)
     {
         fprintf(stderr, "Error: Could not initialize action_block_t\n");
-        return NULL;    
+        return NULL;
     }
-    
-    block_t* block = malloc(sizeof(block));
+
+    block_t *block = malloc(sizeof(block));
```
Comment (sabinecodes15) [inline src/custom-actions/src/action_block.c:73]: Check if malloc is successful

File: src/custom-actions/include/custom-actions-effect.h [hunk 4]
```diff
@@ -131,9 +125,7 @@ int gen_attrval(int min, int max, attribute_t *a);
  *
```
Comment (sabinecodes15) [inline src/custom-actions/include/custom-actions-effect.h]: Update the block comment so it properly reflects the new move_player function. Make sure it's distinct from the comments for `say_phrase`, and update the parameter list.

File: src/custom-actions/src/custom-actions-effect.c [hunk 5]
```diff
@@ -92,111 +92,115 @@ void double_arithmetic(attribute_t *a1, attribute_t *a2, attribute_t *a3,
...
 
+// already implemented in game.h
+/* See custom-actions-effect.h */
+// int move_player(game_t *game, room_t *new_room)
+// {
+
+// }
 
 /* TO BE IMPLEMENTED:
  *
```
Comment (sabinecodes15) [inline src/custom-actions/src/custom-actions-effect.c]: If `move_player(player_t *p, room_t *r)` is being replaced with `move_player(game_t *game, room_t *new_room)`, remove it from the TO BE IMPLEMENTED section below.

File: src/custom-actions/src/custom-actions-effect.c [hunk 6]
```diff
@@ -92,111 +92,115 @@ void double_arithmetic(attribute_t *a1, attribute_t *a2, attribute_t *a3,
...
+        return SUCCEEDS;
+    case INTEGER:
+        a->attribute_value.int_val = val;
+        return SUCCEEDS;
+    default:
+        return FAILS;
     }
 }
 
+// already implemented in game.h
```
Comment (sabinecodes15) [inline src/custom-actions/src/custom-actions-effect.c]: I don't see a `move_player` function implemented in game.h, but I do see `move_room` function. Is this what you're referring to?

File: src/custom-actions/include/custom-actions-effect.h [hunk 7]
```diff
@@ -131,9 +125,7 @@ int gen_attrval(int min, int max, attribute_t *a);
...
  * Returns
  *  - SUCCEEDS/FAILS int status code
- * 
- * int move_player(player_t *p, room_t *r);
  */
-
+int move_player(game_t *game, room_t *r);
```
Comment (sabinecodes15) [inline src/custom-actions/include/custom-actions-effect.h]: Where exactly is this implemented? (See also my comment [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872921347).)

File: tests/custom-actions/test_action_block.c [hunk 8]
```diff
@@ -14,24 +16,28 @@ Test(action_block_t, new_SET)
...
-                 "didn't set new_action->action_type");
+                                                       "didn't set new_action->action_type");
     cr_assert_eq(new_action->num_args, num_args, "action_block_new() didn't "
-                 "set new_action->num_args");
-    cr_assert_eq(new_action->args, &args, "action_block_new() didn't set "
-                 "new_action->args");
-    
+                                                 "set new_action->num_args");
+    cr_assert_eq(new_action->action_params->args, action_params->args, "action_block_new() didn't set "
+                                                                        "new_action->action_params->args");
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c:39]:
```suggestion
                                                                       "new_action->action_params->args");
```

File: tests/custom-actions/test_action_block.c [hunk 9]
```diff
@@ -730,30 +828,36 @@ Test(exec_action_block, set)
...
+
     attr3 = malloc(sizeof(attribute_t));
     attr3->attribute_key = "attr3";
     attr3->attribute_tag = CHARACTER;
     val3.char_val = 'b';
     attr3->attribute_value = val3;
- 
-    args1 = (attribute_t**) malloc(sizeof(attribute_t*) * 2);
+
+    args1 = (attribute_t **)malloc(sizeof(attribute_t *) * 2);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args1 = (attribute_t**)malloc(sizeof(attribute_t*) * 2);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: [Style guide](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872921347) says no space between variable and dereference operator; I think that applies here

File: tests/custom-actions/test_action_block.c [hunk 10]
```diff
@@ -730,30 +828,36 @@ Test(exec_action_block, set)
...
     args1[0] = attr1;
     args1[1] = attr2;
-    
-    args2 = (attribute_t**) malloc(sizeof(attribute_t*) * 2);
+
+    args2 = (attribute_t **)malloc(sizeof(attribute_t *) * 2);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args2 = (attribute_t**)malloc(sizeof(attribute_t*) * 2);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: See [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077)

File: tests/custom-actions/test_action_block.c [hunk 11]
```diff
@@ -805,20 +909,26 @@ void test_arithmetic_act(int i1, int i2, int exp, action_enum_t op)
     attr4->attribute_tag = INTEGER;
     val4.int_val = 0;
     attr4->attribute_value = val4;
- 
-    args1 = (attribute_t**) malloc(sizeof(attribute_t*) * 3);
+
+    args1 = (attribute_t **)malloc(sizeof(attribute_t *) * 3);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args1 = (attribute_t**)malloc(sizeof(attribute_t*) * 3);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 12]
```diff
@@ -805,20 +909,26 @@ void test_arithmetic_act(int i1, int i2, int exp, action_enum_t op)
...
     args1[0] = attr1;
     args1[1] = attr2;
     args1[2] = attr4;
-    
-    args2 = (attribute_t**) malloc(sizeof(attribute_t*) * 3);
+
+    args2 = (attribute_t **)malloc(sizeof(attribute_t *) * 3);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args2 = (attribute_t**)malloc(sizeof(attribute_t*) * 3);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 13]
```diff
@@ -730,30 +828,36 @@ Test(exec_action_block, set)
...
+    args2 = (attribute_t **)malloc(sizeof(attribute_t *) * 2);
     args2[0] = attr1;
     args2[1] = attr3;
- 
-    act1 = action_block_new(SET, 2, args1);
-    act2 = action_block_new(SET, 3, args1);
-    act3 = action_block_new(SET, 2, args2);
+
+    action_params1 = (action_param_t *)malloc(sizeof(action_param_t));
+    action_params2 = (action_param_t *)malloc(sizeof(action_param_t));
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    action_params2 = (action_param_t*)malloc(sizeof(action_param_t));
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces for lines 852 and 853 (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 14]
```diff
@@ -805,20 +909,26 @@ void test_arithmetic_act(int i1, int i2, int exp, action_enum_t op)
...
     args2[0] = attr1;
     args2[1] = attr3;
     args2[2] = attr4;
- 
-    act1 = action_block_new(op, 3, args1);
-    act2 = action_block_new(op, 2, args1);
-    act3 = action_block_new(op, 3, args2);
+
+    action_params1 = (action_param_t *)malloc(sizeof(action_param_t));
+    action_params2 = (action_param_t *)malloc(sizeof(action_param_t));
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    action_params2 = (action_param_t*)malloc(sizeof(action_param_t));
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces for lines 923 and 924 (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 15]
```diff
@@ -907,35 +1013,96 @@ Test(exec_action_block, gen)
     attr4->attribute_tag = INTEGER;
     val4.int_val = 0;
     attr4->attribute_value = val4;
- 
-    args1 = (attribute_t**) malloc(sizeof(attribute_t*) * 3);
+
+    args1 = (attribute_t **)malloc(sizeof(attribute_t *) * 3);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args1 = (attribute_t**)malloc(sizeof(attribute_t*) * 3);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 16]
```diff
@@ -907,35 +1013,96 @@ Test(exec_action_block, gen)
...
     args1[0] = attr1;
     args1[1] = attr2;
     args1[2] = attr4;
-    
-    args2 = (attribute_t**) malloc(sizeof(attribute_t*) * 3);
+
+    args2 = (attribute_t **)malloc(sizeof(attribute_t *) * 3);
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    args2 = (attribute_t**)malloc(sizeof(attribute_t*) * 3);
```

Comment (sabinecodes15) [reply tests/custom-actions/test_action_block.c]: Spaces (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 17]
```diff
@@ -907,35 +1013,96 @@ Test(exec_action_block, gen)
...
     args2[1] = attr3;
     args2[2] = attr4;
- 
-    act1 = action_block_new(GEN, 3, args1);
-    act2 = action_block_new(GEN, 2, args1);
-    act3 = action_block_new(GEN, 3, args2);
-    
+
+    action_params1 = (action_param_t *)malloc(sizeof(action_param_t));
+    action_params2 = (action_param_t *)malloc(sizeof(action_param_t));
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
    action_params2 = (action_param_t*)malloc(sizeof(action_param_t));
```
Spaces for lines 1027 and 1028 (see [here](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872989077))

File: tests/custom-actions/test_action_block.c [hunk 18]
```diff
@@ -907,35 +1013,96 @@ Test(exec_action_block, gen)
...
+    rc3 = exec_action_block(act3);
+
+    cr_assert_eq(rc1, SUCCESS, "Expected SUCCESS but exec_action_block "
+                               "returned FAILURE with action type GEN");
+
+    cr_assert_eq(rc2, FAILURE, "exec_action_block() failed to recognize "
+                               "invalid number of arguments");
+
+    cr_assert_eq(rc1, SUCCESS, "Expected SUCCESS but exec_action_block "
+                               "returned FAILURE with action type GEN");
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
                               "returned FAILURE with action type TELEPORT");
```

File: tests/custom-actions/test_action_block.c [hunk 19]
```diff
@@ -907,35 +1013,96 @@ Test(exec_action_block, gen)
...
+    act1 = action_block_new(TELEPORT, 2, action_params1);
+    act2 = action_block_new(TELEPORT, 1, action_params1);
+    act3 = action_block_new(TELEPORT, 2, action_params2);
+
+    rc1 = exec_action_block(act1);
+    rc2 = exec_action_block(act2);
+    rc3 = exec_action_block(act3);
+
+    cr_assert_eq(rc1, SUCCESS, "Expected SUCCESS but exec_action_block "
+                               "returned FAILURE with action type GEN");
```
Comment (sabinecodes15) [inline tests/custom-actions/test_action_block.c]:
```suggestion
                               "returned FAILURE with action type TELEPORT");
```

**Review by sabinecodes15 — CHANGES_REQUESTED** [review]
[This comment](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872921516) (with this [related comment](https://github.com/uchicago-cs/chiventure/pull/1373/files#r872921347)) is the main issue I noticed; if this feature is working as intended, consider making clearer comments

**Review by ecoble — APPROVED** [review]
Great job on this! Also great job @sabinecodes15 for such a thorough first review! Approving this pull request, as a next step consider writing a small example using teleportation (which could be useful for the demo), or adding teleportation specific tests.

**Review by sabinecodes15 — APPROVED** [review]
Looks good to merge!

**Pull request closed without merging**